nemo/collections/nlp/models/dialogue/intent_bio_type_classification_model.py

```diff
@@ -0,0 +1,793 @@
+        """
+        Update data directory and get data stats with Data Descriptor.
+        Also, reconfigures the classifier - to cope with data with e.g. different number of slots.
+
+        Args:
+            data_dir: path to data directory
+        """
+        logging.info(f'Setting data_dir to {data_dir}.')
+        self.data_dir = data_dir
+        # Update configuration with new data.
+        self._set_data_desc_to_cfg(self.cfg, data_dir, train_ds, validation_ds)
+        # Reconfigure the classifier for different settings (number of intents, slots etc.).
+        self._reconfigure_classifier()
+
+    def update_data_dir_for_testing(self, data_dir) -> None:
+        """
+        Update data directory.
+
+        Args:
+            data_dir: path to data directory
+        """
+        logging.info(f'Setting data_dir to {data_dir}.')
+        self.data_dir = data_dir
+
+    @typecheck()
+    def forward(self, input_ids, attention_mask, token_type_ids, mention_mask):
+        """
+        No special modification required for Lightning, define it as you normally would
+        in the `nn.Module` in vanilla PyTorch.
+        """
+        if self._cfg.tokenizer.get('library', '') == 'megatron':
+            hidden_states, _ = self.bert_model(input_ids, attention_mask, tokentype_ids=token_type_ids, lm_labels=None)
+        else:
+            hidden_states = self.bert_model(
+                input_ids=input_ids, token_type_ids=token_type_ids, attention_mask=attention_mask
+            )
+
+        from os.path import exists
+        global UNIQUE_EPOCH_COUNTER
+
+        # print("START DUMP hidden_states")
+        # path_to_file="/home/lilee/pickle/hidden_states"+str(UNIQUE_EPOCH_COUNTER)+".pickle"
+        # if exists(path_to_file)==False:
+        #     with open(path_to_file, 'wb') as f:
+        #         # Pickle the 'data' dictionary using the highest protocol available.
+        #         pickle.dump(hidden_states, f, pickle.HIGHEST_PROTOCOL)        
+        # print("END DUMP hidden_states")
+
+
+        # intent_logits, slot_logits = self.classifier(hidden_states=hidden_states)
+        intent_logits, bio_slot_logits = self.classifier(hidden_states=hidden_states)
+
+        # print("START DUMP bio_slot_logits")
+        # path_to_file="/home/lilee/pickle/bio_slot_logits"+str(UNIQUE_EPOCH_COUNTER)+".pickle"
+        # if exists(path_to_file)==False:
+        #     with open(path_to_file, 'wb') as f:
+        #         # Pickle the 'data' dictionary using the highest protocol available.
+        #         pickle.dump(bio_slot_logits, f, pickle.HIGHEST_PROTOCOL)
+        # print("END DUMP bio_slot_logits")
+
+        # current:bug; hidden_states#(batch*128*768) zzz
+        hidden_states_dim = hidden_states.size()[2]
+        max_token_len = hidden_states.size()[1]
+
+        mention_hidden_states=[]
+
+        for m, h in zip(mention_mask, hidden_states):
```

Comment:
Can you abstract this out into a method? This makes it much more readable and testable/debuggable. Also can you use more informative variable names? (i.e. one_mention_mask instead of m, one_hidden_state instead of h). Furthermore, it seems like you're not using tensor operations (i.e. torch. ...), this is not only slow but might cause the problem that some operations to be non-differentiable